src/constraints/ObjectConstrains.ts

```diff
@@ -0,0 +1,53 @@
+import get from 'lodash/get';
+import { ExpectedConstraintError } from '../lib/errors/ExpectedConstraintError';
+import { Result } from '../lib/Result';
+import type { BaseValidator } from '../type-exports';
+import type { IConstraint } from './type-exports';
+
+export type ObjectConstraintName = `s.object(T.when)`;
+
+export type WhenKey = PropertyKey | PropertyKey[];
+
+export interface WhenOptions<T extends BaseValidator<any>, Key extends WhenKey> {
+	is?: boolean | ((value: Key extends Array<any> ? any[] : any) => boolean);
+	then: (predicate: T) => T;
+	otherwise?: (predicate: T) => T;
+}
+
+export function whenConstraint<T extends BaseValidator<any>, I, Key extends WhenKey>(
+	key: Key,
+	options: WhenOptions<T, Key>,
+	validator: T
+): IConstraint<I> {
+	return {
+		run(input: I, parent?: any) {
+			if (!parent) {
+				return Result.err(new ExpectedConstraintError('s.object(T.when)', 'Validator has no parent', parent, 'Validator to have a parent'));
+			}
+
+			const isKeyArray = Array.isArray(key);
+
+			const value = isKeyArray ? key.map((k) => get(parent, k)) : get(parent, key);
+
+			const predicate = resolveBooleanIs<T, Key>(options, value, isKeyArray) ? options.then : options.otherwise;
+
+			if (predicate) {
+				return predicate(validator).run(input) as Result<I, ExpectedConstraintError<I>>;
+			}
+
+			return Result.ok(input);
```

Comment:
Should it return the input if no predicate is found?

Reply:
if `otherwise` isn't present then it returns the input.